src/device_handle/async_api.rs

```diff
@@ -0,0 +1,191 @@
+    closure: F,
+    _pin: PhantomPinned, // `ptr` holds a ptr to `closure`, so mark !Unpin
+    _device: PhantomData<&'d DeviceHandle<C>>,
+    _buf: PhantomData<&'b mut [u8]>,
+}
+// TODO: should CbResult lifetime be different from 'b?
+impl<'d, 'b, C: 'd + UsbContext, F: FnMut(CbResult<'b>) + Send> AsyncTransfer<'d, 'b, C, F> {
+    #[allow(unused)]
+    pub fn new_bulk(
+        device: &'d DeviceHandle<C>,
+        endpoint: u8,
+        buffer: &'b mut [u8],
+        callback: F,
+        timeout: std::time::Duration,
+    ) -> Pin<Box<Self>> {
+        // non-isochronous endpoints (e.g. control, bulk, interrupt) specify a value of 0
+        // This is step 1 of async API
+        let ptr = unsafe { ffi::libusb_alloc_transfer(0) };
+        let ptr = NonNull::new(ptr).expect("Could not allocate transfer!");
+        let timeout = libc::c_uint::try_from(timeout.as_millis())
+            .expect("Duration was too long to fit into a c_uint");
+
+        // Safety: Pinning `result` ensures it doesn't move, but we know that we will
+        // want to access its fields mutably, we just don't want its memory location
+        // changing (or its fields moving!). So routinely we will unsafely interact with
+        // its fields mutably through a shared reference, but this is still sound.
+        let result = Box::pin(Self {
+            ptr,
+            closure: callback,
+            _pin: PhantomPinned,
+            _device: PhantomData,
+            _buf: PhantomData,
+        });
+
+        unsafe {
+            // This casting, and passing it to the transfer struct, relies on
+            // the pointer being a regular pointer and not a fat pointer.
+            // Also, closure will be invoked from whatever thread polls, which
+            // may be different from the current thread. So it must be `Send`.
+            // Also, although many threads at once may poll concurrently, only
+            // one will actually ever execute the transfer at a time, so we do
+            // not need to worry about simultaneous writes to the buffer
+            let closure_as_ptr: *mut F = {
+                let ptr: *const F = &result.closure;
+                ptr as *mut F
+            };
+            // Step 2 of async api
+            ffi::libusb_fill_bulk_transfer(
+                ptr.as_ptr(),
+                device.as_raw(),
+                endpoint,
+                buffer.as_ptr() as *mut u8,
+                buffer.len().try_into().unwrap(),
+                Self::transfer_cb,
+                closure_as_ptr.cast(),
+                timeout,
+            )
+        };
+        result
+    }
+
+    /// Submits a transfer to libusb's event engine.
+    /// # Panics
+    /// A transfer should not be double-submitted! Only re-submit after a submission has
+    /// returned Err, or the callback has gotten an Err.
+    // Step 3 of async API
+    #[allow(unused)]
+    pub fn submit(&mut self) -> Result<(), TransferError> {
```

Comment:
Yeah, you're right. I put it as mut because I wasnt sure if it would be safe to have it as a shared reference. Can I just change it to &? Would I need to make it !Sync?

Reply:
Or perhaps it should be `submit(self: &mut Pin<Box<Self>>)` ?

Reply:
I think its fixed now, lmk

Reply:
`&self` may actually be safe given the locking inside `libusb_submit_transfer`, but I'd have to look more closely.